ArduPlane/soaring.cpp

```diff
@@ -66,156 +66,118 @@ void Plane::update_soaring() {
+        if (!ahrs.get_relative_position_NED_home(position)) {
+            return;
+        }
+
+        // Check distance to home against MAX_RADIUS.
+        if (g2.soaring_controller.max_radius >= 0 &&
+            powf(position.x,2)+powf(position.y,2) > powf(g2.soaring_controller.max_radius,2) &&
+            previous_mode->mode_number()!=Mode::Number::AUTO) {
+            // Some other loiter status, and outside of maximum soaring radius, and previous mode wasn't AUTO
+            gcs().send_text(MAV_SEVERITY_INFO, "Soaring: Outside SOAR_MAX_RADIUS, RTL");
+            set_mode(mode_rtl, ModeReason::SOARING_DRIFT_EXCEEDED);
+            break;
+        }
+
+        // If previous mode was AUTO and there was a previous NAV command, we can use previous and next wps for drift calculation
+        // with respect to the desired direction of travel. If these vectors are zero, drift will be calculated from thermal start
+        // position only, without taking account of the desired direction of travel.
         Vector2f prev_wp, next_wp;
 
-        // If previous mode was AUTO and there was a previous NAV command, we can use previous and next wps for drift calculation.
-        if (previous_mode->mode_number() == Mode::Number::AUTO) {
+        if (previous_mode == &mode_auto) {
             AP_Mission::Mission_Command current_nav_cmd = mission.get_current_nav_cmd();
             AP_Mission::Mission_Command prev_nav_cmd;
 
             if (!(mission.get_next_nav_cmd(mission.get_prev_nav_cmd_with_wp_index(), prev_nav_cmd) &&
-                   prev_nav_cmd.content.location.get_vector_xy_from_origin_NE(prev_wp) &&
+                prev_nav_cmd.content.location.get_vector_xy_from_origin_NE(prev_wp) &&
                 current_nav_cmd.content.location.get_vector_xy_from_origin_NE(next_wp))) {
-                prev_wp.x = 0.0;
-                prev_wp.y = 0.0;
-                next_wp.x = 0.0;
-                next_wp.y = 0.0;
+                prev_wp.zero();
+                next_wp.zero();
             }
         }
 
+        // Get the status of the soaring controller cruise checks.
         const SoaringController::LoiterStatus loiterStatus = g2.soaring_controller.check_cruise_criteria(prev_wp/100, next_wp/100);
 
         if (loiterStatus == SoaringController::LoiterStatus::GOOD_TO_KEEP_LOITERING) {
             // Reset loiter angle, so that the loiter exit heading criteria
             // only starts expanding when we're ready to exit.
             plane.loiter.sum_cd = 0;
-            plane.soaring_mode_timer = AP_HAL::millis();
+            plane.soaring_mode_timer_ms = AP_HAL::millis();
+
+            //update the wp location
+            g2.soaring_controller.get_target(next_WP_loc);
+
             break;
         }
 
         // Some other loiter status, we need to think about exiting loiter.
-        uint32_t timer = AP_HAL::millis() - plane.soaring_mode_timer;
+        uint32_t timer_ms = AP_HAL::millis() - plane.soaring_mode_timer_ms;
```

Comment:
This could be `const`.

Reply:
It could also have a better name - `time_spent_in_loiter_ms` or `time_in_loiter_ms` or similar.

Reply:
Done